source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h

```diff
@@ -15,20 +18,27 @@ namespace LocalRateLimit {
 
 class LocalRateLimiterImpl {
 public:
-  LocalRateLimiterImpl(const std::chrono::milliseconds fill_interval, const uint32_t max_tokens,
-                       const uint32_t tokens_per_fill, Event::Dispatcher& dispatcher);
+  LocalRateLimiterImpl(
+      const std::chrono::milliseconds fill_interval, const uint32_t max_tokens,
+      const uint32_t tokens_per_fill, Event::Dispatcher& dispatcher,
+      const Protobuf::RepeatedPtrField<
+          envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>& descriptors);
   ~LocalRateLimiterImpl();
 
-  bool requestAllowed() const;
+  bool requestAllowed(const std::vector<RateLimit::LocalDescriptor>& request_descriptors) const;
 
 private:
   void onFillTimer();
+  void onFillTimerHelper(const RateLimit::TokenState& state, const RateLimit::TokenBucket& bucket);
+  void onFillTimerDescriptorHelper();
+  bool requestAllowedHelper(const RateLimit::TokenState& tokens) const;
 
-  const std::chrono::milliseconds fill_interval_;
-  const uint32_t max_tokens_;
-  const uint32_t tokens_per_fill_;
+  RateLimit::TokenBucket token_bucket_;
   const Event::TimerPtr fill_timer_;
-  mutable std::atomic<uint32_t> tokens_;
+  TimeSource& time_source_;
+  RateLimit::TokenState tokens_;
+  absl::flat_hash_set<RateLimit::LocalDescriptor> descriptors_;
+  std::vector<std::unique_ptr<RateLimit::TokenState>> descriptor_tokens_;
```

Comment:
Per my above comment, I'm wondering if you could have a LocalDescriptorImpl which stores the token state and then you don't need this other vector?

Reply:
Yeah, will switch to heterogeneous lookups instead.